kvcoord: add 'have been waiting' warning to DistSender

tbg opened the pull request:
We already have these at the Range level, but it looks like we might
have seen a case in which requests spent all of their time in
DistSender, without ever showing up on the ranges. This could be an
infinite retry loop related to descriptor caching, though this is pure
speculation. Either way, it will be helpful to see what error these
requests are stuck retrying on should the problem occur again.

No release note since this is out in the weeds.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/47264)
<!-- Reviewable:end -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47264#-:-M4UpQRH1lkxpcfKsYuS:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/kv/kvclient/kvcoord/dist_sender.go, line 1421 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47264#-M4UpEOO6549eyERCV58:-M4UpEOP1E-7mqHyErB2:b-xgdmbf) ([raw file](https://github.com/cockroachdb/cockroach/blob/b65ae07e1416abf33b0b50cdfe2a4e5de5577aac/pkg/kv/kvclient/kvcoord/dist_sender.go#L1421)):*
> ```Go
> 		if dur := timeutil.Since(tBegin); dur > slowDistSenderThreshold && attempts >= 0 {
> 			log.Warning(ctx, slowRangeRPCWarningStr(dur, attempts, desc, pErr))
> 			attempts = math.MinInt64 // avoid logging over and over for this request
> ```

Consider implementing this using `log.Every` instead so that we do continue to hear from the request periodically.


<!-- Sent from Reviewable.io -->

tbg commented:
I gave this the 'defer' treatment (which is what we use for all other instances of this kind of log msg), PTAL. I also added the gauge we usually have for these.

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47264#-:-M5NHsou3e01bJvKAijN:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/kv/kvclient/kvcoord/dist_sender.go, line 105 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47264#-M5NHvxd4gel-cYUPDkr:-M5NHvxd4gel-cYUPDks:b-i2bku7) ([raw file](https://github.com/cockroachdb/cockroach/blob/ade4f36ef257fff6aa0441a42197013a23881032/pkg/kv/kvclient/kvcoord/dist_sender.go#L105)):*
> ```Go
> 	metaDistSenderSlowRPCs = metric.Metadata{
> 		Name:        "requests.slow.distsender",
> 		Help:        "Number of range RPCs that have been stuck or retrying for a long time",
> ```

What is a "range RPC"? Given the help text for `distsender.rpc.sent`, I'd expect this to just be "Number of RPCs ...".
___
*[pkg/kv/kvclient/kvcoord/dist_sender_test.go, line 3491 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47264#-M5NHoDt9Yva68rEq9hn:-M5NHoDt9Yva68rEq9ho:bul0cqf) ([raw file](https://github.com/cockroachdb/cockroach/blob/ade4f36ef257fff6aa0441a42197013a23881032/pkg/kv/kvclient/kvcoord/dist_sender_test.go#L3491)):*
> ```Go
> 	desc := &roachpb.RangeDescriptor{RangeID: 9, StartKey: roachpb.RKey("x")}
> 	{
> 		act := slowRangeRPCWarningStr(
> ```

nit: `act` and `exp` are flipped in the two test cases.


<!-- Sent from Reviewable.io -->

tbg committed:
kvcoord: add 'have been waiting' warning to DistSender

We already have these at the Range level, but it looks like we might
have seen a case in which requests spent all of their time in
DistSender, without ever showing up on the ranges. This could be an
infinite retry loop related to descriptor caching, though this is pure
speculation. Either way, it will be helpful to see what error these
requests are stuck retrying on should the problem occur again.

No release note since this is out in the weeds.

Release note: None

cockroach-teamcity commented:
[![CLA assistant check](https://cla.crdb.dev/pull/badge/not_signed)](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=47264) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=47264) before we can accept your contribution.<br/>**11** out of **12** committers have signed the CLA.<br/><br/>:white_check_mark: otan<br/>:white_check_mark: andreimatei<br/>:white_check_mark: yuzefovich<br/>:white_check_mark: RaduBerinde<br/>:white_check_mark: tbg<br/>:white_check_mark: nvanbenschoten<br/>:white_check_mark: knz<br/>:white_check_mark: asubiotto<br/>:white_check_mark: mjibson<br/>:white_check_mark: rohany<br/>:white_check_mark: jordanlewis<br/>:x: craig[bot]<br/><hr/>**craig[bot]** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.crdb.dev/check/cockroachdb/cockroach?pullRequest=47264) it.</sub>

tbg commented:
bors r=nvanbenschoten
